### PR TITLE
WIP: Expose support for <iterator> and <array>

### PR DESCRIPTION
**Pull request by wmaxey**
Standard libcu++ feature, will fill in PR template soon.

### CODE REVIEW

**Comment by wmaxey** [comment]
Closes #51

**Comment by wmaxey** [comment]
@griwes / @jrhemstad If anyone wants to take a particular look at changes for `initializer_list` I think the current solution is brutally hacky. Potentially the best we can do is add an implicit constructor for a `cuda::std::initializer_list`, but I'm uncertain if we want to add something so confusing. Same issue as `tuple_size` honestly in that we can't go around conflicting with the host standard library.

**Comment by wmaxey** [comment]
VDVS: https://builds4u.nvidia.com/dvs/#/change/3090343059892240.1?showTab=DVS

**Comment by jrhemstad** [comment]
> @griwes / @jrhemstad If anyone wants to take a particular look at changes for `initializer_list` I think the current solution is brutally hacky.

Can you summarize the problem/solution? I'm guessing something to do with `std::initializer_list` gets special treatment from the compiler that a "user-defined" `cuda::std::initializer_list` won't get?

File: include/cuda/std/array [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+#define _CUDA_ARRAY
+
+#include "cassert"
+#include "cstdint"
+#include "limits"
+#include "type_traits"
+#include "iterator"
+#include "utility"
+
+#include <initializer_list>
```
Comment (wmaxey) [inline include/cuda/std/array:19]: @jrhemstad This is mostly what I was talking about. We can't use `cuda::std::initializer_list`, due to the implicit construction of a `std::initializer_list`.

The compiler might insert something like the below, and if we are really certain on using a `cuda::std::initializer_list` we would need to add an implicit conversion which makes 'our' `initializer_list` behave weird: https://gcc.godbolt.org/z/EG739eEdP

http://eel.is/c++draft/dcl.init.list#5

File: include/cuda/std/detail/libcxx/include/iterator [hunk 2]
```diff
@@ -1749,14 +1749,14 @@ reverse_iterator<_Tp*> rend(_Tp (&__array)[_Np])
 
 template <class _Ep>
 _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR_AFTER_CXX14
-reverse_iterator<const _Ep*> rbegin(initializer_list<_Ep> __il)
+reverse_iterator<const _Ep*> rbegin(::std::initializer_list<_Ep> __il)
```
Comment (wmaxey) [inline include/cuda/std/detail/libcxx/include/iterator:1752]: @jrhemstad hacks here.

File: include/cuda/std/array (hunk 1 above)
Comment (jrhemstad) [reply include/cuda/std/array:19]: Yeah, I don't see how we're going to get around this. `intializer_list` is a type that has special rules that the compiler follows that we can't really emulate with `cuda::std::initializer_list`.

File: include/cuda/std/initializer_list [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+#include "detail/__config"
+
+#include <initializer_list>
+
+namespace cuda {
+  namespace std {
+    using ::std::initializer_list;
+  }
+}
```
Comment (jrhemstad) [inline include/cuda/std/initializer_list:20]: Should be like this, right? Otherwise it misses the inline namespace? 
```suggestion
_LIBCUDACXX_BEGIN_NAMESPACE_STD
    using ::std::initializer_list;
_LIBCUDACXX_END_NAMESPACE_STD

```

Comment (wmaxey) [reply include/cuda/std/initializer_list:20]: Not sure!

I think it's fine to include the ABI as the namespace include will still resolve to a 'std::' thing. I'll make the switch in another patch just since it's consistent.

File: .upstream-tests/maintenance/std-to-cuda [hunk 4]
```diff
@@ -7,7 +7,7 @@ root_dir=$(cd "$(dirname "${BASH_SOURCE[0]}")/.." && pwd)
 stdlib_headers=$(<"${root_dir}/maintenance/stdlib-headers")
 header_replacements=$(echo "${stdlib_headers}" | sed 's#<\(.*\)>#-e s:<\1>:<cuda/std/\1>:g#')
 
-find "${root_dir}/test" -name "*.cpp" |
+find "${root_dir}/test/std/iterators" -name "*.cpp" |
```
Comment (griwes) [inline .upstream-tests/maintenance/std-to-cuda:10]: I don't think you meant to keep this ;>

File: include/cuda/std/detail/libcxx/include/algorithm [hunk 5]
```diff
@@ -13,7 +13,7 @@
 /*
     algorithm synopsis
 
-#include <initializer_list>
+#include <::std::initializer_list>
```
Comment (griwes) [inline include/cuda/std/detail/libcxx/include/algorithm:16]: Tell me you've just run search+replace without telling me... :D

File: include/cuda/std/detail/libcxx/include/algorithm [hunk 6]
```diff
@@ -3302,6 +3336,7 @@ partition_copy(_InputIterator __first, _InputIterator __last,
 // partition_point
 
 template<class _ForwardIterator, class _Predicate>
+_LIBCUDACXX_INLINE_VISIBILITY
```
Comment (griwes) [inline include/cuda/std/detail/libcxx/include/algorithm:3339]: We probably also want `_LIBCUDACXX_EXECUTION_SPACE_SPECIFIER` for the ones here and above? I assume this is where you invented the new macro.

File: include/cuda/std/detail/libcxx/include/array [hunk 7]
```diff
@@ -137,19 +139,19 @@ struct _LIBCUDACXX_TEMPLATE_VIS array
...
-    typedef std::reverse_iterator<const_iterator> const_reverse_iterator;
+    typedef std::reverse_iterator<const_iterator>      const_reverse_iterator;
+    typedef std::reverse_iterator<iterator>            reverse_iterator;
 
     _Tp __elems_[_Size];
 
     // No explicit construct/copy/destroy for aggregate type
     _LIBCUDACXX_INLINE_VISIBILITY void fill(const value_type& __u) {
-      _CUDA_VSTD::fill_n(__elems_, _Size, __u);
+      fill_n(__elems_, _Size, __u);
```
Comment (griwes) [inline include/cuda/std/detail/libcxx/include/array:149]: This (and swap_ranges below) are now ADL calls, we need to keep them qualified.

File: include/cuda/std/detail/libcxx/include/iterator [hunk 8]
```diff
@@ -714,7 +714,7 @@ public:
     _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR_AFTER_CXX14
     reference operator*() const {_Iter __tmp = current; return *--__tmp;}
     _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR_AFTER_CXX14
-    pointer  operator->() const {return _CUDA_VSTD::addressof(operator*());}
+    pointer  operator->() const {return addressof(operator*());}
```
Comment (griwes) [inline include/cuda/std/detail/libcxx/include/iterator:717]: Needs to be qualified to avoid ADL.

File: include/cuda/std/detail/libcxx/include/iterator [hunk 9]
```diff
@@ -1168,7 +1168,7 @@ public:
     _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR_AFTER_CXX14
     move_iterator& operator-=(difference_type __n) {__i -= __n; return *this;}
     _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR_AFTER_CXX14
-    reference operator[](difference_type __n) const { return static_cast<reference>(__i[__n]); }
+    reference operator[](difference_type __n) const { return move(__i[__n]); }
```
Comment (griwes) [inline include/cuda/std/detail/libcxx/include/iterator:1171]: Ditto.

File: include/cuda/std/detail/libcxx/include/iterator (hunk 2 above)
Comment (griwes) [reply include/cuda/std/detail/libcxx/include/iterator:1752]: This seems fine to me.

File: include/cuda/std/detail/libcxx/include/stdexcept [hunk 10]
```diff
@@ -229,7 +231,11 @@ void __throw_logic_error(const char*__msg)
     throw logic_error(__msg);
 #else
     ((void)__msg);
-    _CUDA_VSTD::abort();
+    NV_IF_ELSE_TARGET(
+      NV_IS_DEVICE,
+        __trap();,
+        _CUDA_VSTD::abort();
+    )
```
Comment (griwes) [inline include/cuda/std/detail/libcxx/include/stdexcept:238]: Somewhere above you're made a macro for this, though using if/else on `__CUDA_ARCH__` instead of if target. Should we replace that with what's in here, and just use the macro here to avoid repeating this construct?

**Review by griwes — CHANGES_REQUESTED** [review]

File: include/cuda/std/detail/libcxx/include/algorithm (hunk 6 above)
Comment (wmaxey) [reply include/cuda/std/detail/libcxx/include/algorithm:3339]: I'm wary to do that, when building source files in libcxx there are places where some extra ABI flags are set that may matter. The ones I've marked are mainly internal and inline.

Putting `_LIBCUDACXX_INLINE_VISIBLITY` on those causes them to fail to build due to the linkage flags.

e.g.
```
#if __has_attribute(internal_linkage)
#  define _LIBCUDACXX_INTERNAL_LINKAGE __attribute__ ((internal_linkage))
#else
#  define _LIBCUDACXX_INTERNAL_LINKAGE _LIBCUDACXX_ALWAYS_INLINE
#endif
```

File: include/cuda/std/detail/libcxx/include/stdexcept (hunk 10 above)
Comment (wmaxey) [reply include/cuda/std/detail/libcxx/include/stdexcept:238]: Which macro was this?

File: .upstream-tests/maintenance/std-to-cuda (hunk 4 above)
Comment (wmaxey) [reply .upstream-tests/maintenance/std-to-cuda]: This should be fixed now.

File: include/cuda/std/detail/libcxx/include/algorithm (hunk 5 above)
Comment (wmaxey) [reply include/cuda/std/detail/libcxx/include/algorithm]: Also fixed. >.>

File: include/cuda/std/detail/libcxx/include/array (hunk 7 above)
Comment (wmaxey) [reply include/cuda/std/detail/libcxx/include/array]: I've replaced `_CUDA_VSTD::` on this and below.

**Review by griwes — APPROVED** [review]
